x-pack/plugins/ingest_manager/server/services/epm/elasticsearch/transform/install.ts

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObjectsClientContract } from 'kibana/server';
+
+import { saveInstalledEsRefs } from '../../packages/install';
+import * as Registry from '../../registry';
+import {
+  Dataset,
+  ElasticsearchAssetType,
+  EsAssetReference,
+  RegistryPackage,
+} from '../../../../../common/types/models';
+import { CallESAsCurrentUser } from '../../../../types';
+import { getInstallation } from '../../packages';
+import { deleteTransforms, deleteTransformRefs } from './remove';
+import { getAsset } from './common';
+
+interface TransformInstallation {
+  installationName: string;
+  content: string;
+}
+
+interface TransformPathDataset {
+  path: string;
+  dataset: Dataset;
+}
+
+export const installTransformForDataset = async (
+  registryPackage: RegistryPackage,
+  paths: string[],
+  callCluster: CallESAsCurrentUser,
+  savedObjectsClient: SavedObjectsClientContract
+) => {
+  const installation = await getInstallation({ savedObjectsClient, pkgName: registryPackage.name });
+  let previousInstalledTransformEsAssets: EsAssetReference[] = [];
+  if (installation) {
+    previousInstalledTransformEsAssets = installation.installed_es.filter(
+      ({ type, id }) => type === ElasticsearchAssetType.transform
+    );
+  }
+
+  // delete all previous transform
+  await deleteTransforms(
+    callCluster,
+    previousInstalledTransformEsAssets.map((asset) => asset.id)
+  );
+  // install the latest dataset
+  const datasets = registryPackage.datasets;
+  if (!datasets?.length) return [];
+  const installNameSuffix = `${registryPackage.version}`;
+
+  const transformPaths = paths.filter((path) => isTransform(path));
+
+  const transformPathDatasets = datasets.reduce<TransformPathDataset[]>((acc, dataset) => {
+    transformPaths.forEach((path) => {
+      if (isDatasetTransform(path, dataset.path)) {
+        acc.push({ path, dataset });
+      }
+    });
+    return acc;
+  }, []);
+
+  const transformRefs = transformPathDatasets.reduce<EsAssetReference[]>(
+    (acc, transformPathDataset) => {
+      if (transformPathDataset) {
+        acc.push({
+          id: getTransformNameForInstallation(transformPathDataset, installNameSuffix),
+          type: ElasticsearchAssetType.transform,
+        });
+      }
+      return acc;
+    },
+    []
+  );
+
+  // get and save transform refs before installing transforms
+  await saveInstalledEsRefs(savedObjectsClient, registryPackage.name, transformRefs);
```

Comment:
If there is no `transformPaths` there won't be any `transformRefs` so no need to update the saved object.